### PR TITLE
Use KubernetesHook to create api client in KubernetesPodOperator

### PR DESCRIPTION
**Pull request by dstandish**
Reopened and ready for review.

TLDR: add support for k8s hook in KPO; use it always (even when no conn id); continue to consider the core k8s settings that KPO already takes into account but emit deprecation warning about them.

This one was pretty tricky.

Here are some notes:

KPO takes into account a few settings from core airflow cfg (e.g. verify ssl, tcp keepalive, context, config file, and in_cluster).  So if we're using the hook to generate the client, somehow the hook has to take these settings into account. But we don't want the hook to consider these settings in general.  

So we need to read the settings in KPO and somehow pass them to the hook.  We can't do it through init params, because that would interfere with migration to an airflow connection.  So what I did was establish some "private" attribtues that could be patched from KPO when the setting is something other than the default.  Then the hook will consider these patched params only when it's not using an airflow connection or hook params.

The other tricky part was we want to minimize the warnings.  So we need to try and warn only when the configuration taken from core settings actually ends up being used.  E.g. `[kubernetes] > in_cluster=True` is the default value so we only want to warn when it's been set to False, and only when it is actually considered (so that when the user adds an airflow connection, the warning will go away).

Lastly one thing to note (smaller thing), with two settings, when adding them to the hook, I changed the wording.  Core airflow setting `verify_ssl` is True by default.  When adding this to the hook, I add it as `disable_verify_ssl`.  And the reason for this is, `False` seems like a safer default value.  So the default would be `disable_verify_ssl=False`.  E.g. if you did `if extra.get('disable_verify_ssl'): disable` that is safe when the key is not there.  But if you do `if not extra.get('verify_ssl'): disable` well this will disable SSL when the value is not there.  To make this work you'd have to do `if extra.get('verify_ssl') is False`.  The other thing

### CODE REVIEW

File: airflow/providers/cncf/kubernetes/CHANGELOG.rst [hunk 1]
```diff
@@ -103,6 +103,12 @@ Other changes in ``pod_launcher.py``:
 
 * Enum-like class ``PodStatus`` is renamed ``PodPhase``, and the values are no longer lower-cased.
 
+Features
+~~~~~~~~
+
+* Can now derive kubernetes creds using airflow connection, through KubernetesHook
```
Comment (BasPH) [inline airflow/providers/cncf/kubernetes/CHANGELOG.rst]:
```suggestion
* Can now derive Kubernetes credentials from an Airflow connection through the KubernetesHook
```

**Review by BasPH — COMMENTED** [review]
> It doesn't seem that KubernetesHook lookss at these settings. Do you think we need to implement either / or logic here? That is, only use K8s hook if an airflow conn id is provided? And otherwise, continue to use get_kube_client?

Yea that seems reasonable to me

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 2]
```diff
@@ -348,15 +351,17 @@ def launcher(self) -> pod_launcher.PodLauncher:
         return pod_launcher.PodLauncher(kube_client=self.client)
 
     @cached_property
-    def client(self) -> CoreV1Api:
-        # todo: use airflow Connection / hook to authenticate to the cluster
-        kwargs: Dict[str, Any] = dict(
+    def k8s_hook(self):
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]:
```suggestion
    def hook(self):
```

nit, for consistency with other operators?

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 3]
```diff
@@ -348,15 +351,17 @@ def launcher(self) -> pod_launcher.PodLauncher:
...
             config_file=self.config_file,
+            in_cluster=self.in_cluster,
         )
-        if self.in_cluster is not None:
-            kwargs.update(in_cluster=self.in_cluster)
-        return kube_client.get_kube_client(**kwargs)
+
+    @cached_property
+    def client(self) -> CoreV1Api:
+        return self.k8s_hook.core_v1_client
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]:
```suggestion
        return self.hook.core_v1_client
```

**Review by jedcunningham — COMMENTED** [review]
Probably makes sense to do the "or" logic with `get_kube_client`. I do wonder if we should emit a deprecation warning when a connection isn't used?

**Comment by dstandish** [comment]
@jedcunningham @BasPH after some time away from this one I think I see a good way forward.

We can use the `kubernetes` config settings as a "default" for hook behavior.

I.e. we can update the hook so that, it optionally uses these config  values if not overrided by hook params or connection attrs.

```
    if conf.getboolean('kubernetes', 'enable_tcp_keepalive'):
        _enable_tcp_keepalive()

    if not conf.getboolean('kubernetes', 'verify_ssl'):
        _disable_verify_ssl()
```

If these two params are the only thing "missing" ( i.e. in the hook logic compared with the get_kube_client logic ), then it should be fully backward compatible.  Then we can  _always_ use the hook and the parameters (context, config, in cluster) can be freely combined with the hook.

---

Do you have any concerns with using the `kubernetes` config settings as a "default" for hook behavior?  I suppose some could say this integrates provider and core in a way that could potentially be undesirable.  Though others might think it is appropriate to have  the K8s hook take the same defaults as K8s executor.

If we're OK with that, then the question I just need to confirm is, is there anything _else_ that differs between hook and the old get_kube_client

---

@jedcunningham one issue with deprecating use of  those params is, maybe people want to still use  those params and not a conn id.   Use of those params only (i.e. and not also a conn_id) is supported with the hook.  What I might suggest we do though is add a dict param `hook_params` or `hook_kwargs` and then deprecate those three params _as individual params_ in order to reduce number of params in KPO and be more future proof.

**Comment by dstandish** [comment]
I will pick this one back up after merge of https://github.com/apache/airflow/pull/20759

Once that is merged it will be easier / cleaner to have the behavior be consistent between hook and "core" kubernetes approach.

**Comment by dstandish** [comment]
Alright we now have the @potiuk's blessing for this one.  So just need some reviews.  @ashb discussed this one with you a while back.  @raphaelauv I know you are interested in this one.

**Comment by dstandish** [comment]
> could we rename `config_file` -> `config_file_path` in the kubernetes hook ?
> 
> cause currently I find very disturbing the variable name : `config_file`

oy, yeah, you are right, could be named better.  i think i'm the one that added it a couple months back as prep work for this change, and i just kept the same name used in [KPO](https://github.com/apache/airflow/blob/fe6ec34f0d0e7d47d58913a71777fbfb6239b634/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L186).  

But honestly at this point I don't think I'm in favor of changing it.  There's already been so much tumult and change with K8s /  KPO, and really it's not _that_ bad is it?  The risk is what, that a user would think it accepts a file handler?  It's annotated with Optional[str] (though maybe we should extend that to accept `Path`).  And the docstring should also clarify (though i guess i need to add that!)

In any case, if we _do_ want to modify it, I think it belongs in a distinct PR for that.

**Review by potiuk — APPROVED** [review]

File: airflow/providers/cncf/kubernetes/CHANGELOG.rst [hunk 4]
```diff
@@ -19,6 +19,19 @@
...
+
+Features
+~~~~~~~~
+
+KubernetesPodOperator now uses KubernetesHook
+`````````````````````````````````````````````
+
+Previously, KubernetesPodOperator relied on core Airflow configuration (namely setting for kubernetes executor) for certain settings used in client generation.  Now KubernetesPodOperator uses KubernetesHook, and the consideration of core k8s settings is officially deprecated.
+
+If you are using the Airflow configuration settings (e.g. as opposed to operator params) to configure the kubernetes client, then prior to the next major release you will need to add an Airflow connection and set your KPO processes to use that connection.
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/CHANGELOG.rst]:
```suggestion
If you are using the Airflow configuration settings (e.g. as opposed to operator params) to configure the kubernetes client, then prior to the next major release you will need to add an Airflow connection and set your KPO tasks to use that connection.
```

nit

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 5]
```diff
@@ -108,37 +121,77 @@ def __init__(
...
+        return self.conn_extras.get(prefixed_name) or None
+
+    @staticmethod
+    def _deprecation_warning_core_param(deprecation_warnings):
+        settings_list_str = ''.join([f"\n\t{k}={v!r}" for k, v in deprecation_warnings])
+        warnings.warn(
+            f"\nApplying core Airflow settings from section [kubernetes] with the following keys:"
+            f"{settings_list_str}\n"
+            "In a future release, KubernetesPodOperator will no longer consider core\n"
+            "airflow settings; define an Airflow connection instead.",
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py]:
```suggestion
            "Airflow settings; define an Airflow connection instead.",
```

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 6]
```diff
@@ -561,6 +569,39 @@ def dry_run(self) -> None:
...
+        Here we read config from core Airflow config [kubernetes] section.
+        In a future release we will stop looking at this section and require users
+        to use Airflow connections to configure KPO.
+
+        When we find values there that we need to apply on the hook, we patch special
+        hook attributes here.
+        """
+
+        # default for enable_tcp_keepalive is True; patch if False
+        if conf.getboolean('kubernetes', 'enable_tcp_keepalive', fallback=None) is False:
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: I'm curious, why do we need `fallback=None`?

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 7]
```diff
@@ -582,10 +586,12 @@ def test_xcom_push(self, xcom_push):
         self.expected_pod['spec']['containers'].append(container)
         assert self.expected_pod == actual_pod
 
-    @mock.patch("airflow.providers.cncf.kubernetes.utils.pod_manager.PodManager.create_pod")
-    @mock.patch("airflow.providers.cncf.kubernetes.utils.pod_manager.PodManager.await_pod_completion")
-    @mock.patch("airflow.kubernetes.kube_client.get_kube_client")
-    def test_envs_from_secrets(self, mock_client, await_pod_completion_mock, create_pod):
+    @mock.patch(f"{POD_MANAGER_CLASS}.create_pod")
```
Comment (jedcunningham) [inline kubernetes_tests/test_kubernetes_pod_operator.py:639]: 👍

File: tests/providers/cncf/kubernetes/hooks/test_kubernetes.py [hunk 8]
```diff
@@ -131,6 +136,76 @@ def test_get_default_client(
...
+        """
+        Verifies whether disable verify ssl is called depending on combination of hook param and
+        connection extra. Hook param should beat extra.
+        """
+        kubernetes_hook = KubernetesHook(conn_id=conn_id, disable_verify_ssl=disable_verify_ssl)
+        api_conn = kubernetes_hook.get_conn()
+        if disable_called:
+            assert mock_disable.called
+        else:
+            assert not mock_disable.called
```
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/hooks/test_kubernetes.py]:
```suggestion
        assert mock_disable.called is disabled_called
```

nit

**Review by jedcunningham — APPROVED** [review]

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 6 above)
Comment (dstandish) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: if the value is missing from the config and you don't have a fallback, it would throw an error.
as to why None, well, just a reasonable default that means "not set".  Only when it's explicitly set to False would we want to set the hook to disable TCP keepalive.

File: tests/providers/cncf/kubernetes/hooks/test_kubernetes.py (hunk 8 above)
Comment (dstandish) [reply tests/providers/cncf/kubernetes/hooks/test_kubernetes.py]: 🤯 mind blown

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 6 above)
Comment (jedcunningham) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: For those following along at home, _some_ of these fallbacks had no effect (like this one) as conf falls back to the defaults first. However, some of the others have no default ever, and are necessary.

**Review by jedcunningham — APPROVED** [review]

File: tests/providers/cncf/kubernetes/hooks/test_kubernetes.py [hunk 9]
```diff
@@ -131,6 +136,73 @@ def test_get_default_client(
...
+        conn_id,
+        disable_called,
+    ):
+        """
+        Verifies whether disable verify ssl is called depending on combination of hook param and
+        connection extra. Hook param should beat extra.
+        """
+        kubernetes_hook = KubernetesHook(conn_id=conn_id, disable_verify_ssl=disable_verify_ssl)
+        api_conn = kubernetes_hook.get_conn()
+        assert mock_disable.called is disabled_called
```
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/hooks/test_kubernetes.py]:
```suggestion
        assert mock_disable.called is disable_called
```

Oops! Sorry for my typo.

**Review by potiuk — APPROVED** [review]
